includes/class-wc-gateway-stripe.php

```diff
@@ -1021,11 +1021,42 @@ public function get_intent_and_secret() {
 			),
 		);
 
+		/**
+		 * If a saved card has been chosen, use its source object
+		 * for the payment intent in order to let intents handle 3DS.
+		 */
+		if ( isset( $_POST['saved_card'] ) && ! empty( $_POST['saved_card'] ) ) {
+			if ( ! is_user_logged_in() ) {
+				$localized_message = __( 'You need to be logged in to use a saved card.', 'woocommerce-gateway-stripe' );
+				throw new WC_Stripe_Exception( 'saved_card_logged_out', $localized_message );
+			}
+
+			$user_id    = get_current_user_id();
+			$saved_card = intval( $_POST['saved_card'] );
+			$wc_token   = WC_Payment_Tokens::get( $saved_card );
+			$customer   = new WC_Stripe_Customer( $user_id );
+
+			if ( ! $wc_token || $wc_token->get_user_id() !== $user_id ) {
+				throw new WC_Stripe_Exception( 'Invalid payment method', __( 'Invalid payment method. Please input a new card number.', 'woocommerce-gateway-stripe' ) );
+			}
+
+			$source_id = $wc_token->get_token();
+			if ( $this->is_type_legacy_card( $source_id ) ) {
+				// Legacy tokens are not supported in the context of payment intents.
```

Comment:
I think theuser-facing error message should be more explicit. And we should delete the card at this point because it won't ever be able to be used again. Thoughts?

Reply:
After following the logic within `WC_Gateway_Stripe` and `WC_Stripe_Payment_Gateway` closely, I found out that tokens are considered as usual sources by Stripe's API and we can safely use their IDs when generating intents. I'm waiting for Roy to hopefully confirm my theory shortly and I'll update this comment.

Reply:
As suspected, it does work 🎉 

Here is how to test it:

1. Do a sanity check: Without digging too deep in the history of tokens, I see that `WC_Stripe_Payment_Gateway::is_type_legacy_card` uses `/^card_/` to test for a token, based on its usage in `WC_Stripe_Payment_Gateway::prepare_source()`. The following steps are built upon this assumption.
2. Open `https://dashboard.stripe.com/test/customers`. You should already have at least one customer, click it.
3. Scroll down to the *Cards* section and add a new card. Pick one of [these cards](https://stripe.com/docs/testing#three-ds-cards).
4. When you do that and click to edit the card, you should see an ID like `card_1E9vLdIDdSUSdcIrXWLOlWNv`. Copy it for later.
5. Within your dev environment, go to *My Account &raquo; Payment methods* and add the same card.
6. Open the `wp_woocommerce_payment_tokens` database table. You should see something like `src_1E9vfNIDdSUSdcIrMH7wHfU5` in the `token` column. replace it with the card ID from above.
7. Perform a checkout and make sure that the demo 3DS flow is working.